optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/DefaultSolverJob.java

```diff
@@ -0,0 +1,119 @@
+import java.util.function.Function;
+
+import org.optaplanner.core.api.domain.solution.PlanningSolution;
+import org.optaplanner.core.api.solver.Solver;
+import org.optaplanner.core.api.solver.SolverJob;
+import org.optaplanner.core.api.solver.SolverStatus;
+
+/**
+ * @param <Solution_> the solution type, the class with the {@link PlanningSolution} annotation
+ * @param <ProblemId_> the ID type of a submitted problem, such as {@link Long} or {@link UUID}.
+ */
+public class DefaultSolverJob<Solution_, ProblemId_> implements SolverJob<Solution_, ProblemId_>, Callable<Solution_> {
+
+    private final DefaultSolverManager<Solution_, ProblemId_> solverManager;
+    private final ProblemId_ problemId;
+    private final Solver<Solution_> solver;
+    private final Function<ProblemId_, Solution_> problemFinder;
+    private final Consumer<Solution_> finalBestSolutionConsumer;
+    private final BiConsumer<ProblemId_, Throwable> exceptionHandler;
+
+    private volatile SolverStatus solverStatus;
+    private Future<Solution_> future;
+
+    public DefaultSolverJob(DefaultSolverManager<Solution_, ProblemId_> solverManager, ProblemId_ problemId,
+            Solver<Solution_> solver,
+            Function<ProblemId_, Solution_> problemFinder,
+            Consumer<Solution_> finalBestSolutionConsumer,
+            BiConsumer<ProblemId_, Throwable> exceptionHandler) {
+        this.solverManager = solverManager;
+        this.problemId = problemId;
+        this.solver = solver;
+        this.problemFinder = problemFinder;
+        this.finalBestSolutionConsumer = finalBestSolutionConsumer;
+        this.exceptionHandler = exceptionHandler;
+        solverStatus = SolverStatus.SOLVING_SCHEDULED;
+    }
+
+    public void setFuture(Future<Solution_> future) {
+        this.future = future;
+    }
+
+    @Override
+    public ProblemId_ getProblemId() {
+        return problemId;
+    }
+
+    @Override
+    public SolverStatus getSolverStatus() {
+        return solverStatus;
+    }
+
+    @Override
+    public Solution_ call() {
+        solverStatus = SolverStatus.SOLVING_ACTIVE;
+        try {
+            Solution_ problem = problemFinder.apply(problemId);
+            final Solution_ finalBestSolution = solver.solve(problem);
+            if (finalBestSolutionConsumer != null) {
+                // TODO consumption should happen on different thread than solver thread
+                finalBestSolutionConsumer.accept(finalBestSolution);
+            }
+            return finalBestSolution;
+        } catch (Exception e) {
+            exceptionHandler.accept(problemId, e);
+            throw new IllegalStateException("Solving failed for problemId (" + problemId + ").", e);
+        } finally {
+            // TODO What happens if this code throws an exception? Is it eaten?
+            solverManager.getProblemIdToSolverJobMap().remove(problemId);
+            solverStatus = SolverStatus.NOT_SOLVING; // TODO FIXME race condition, lock on problemId
```

Comment:
Sounds like an important `FIXME`. :-)

Reply:
Agreed to postpone from this PR into the next PR for this work that will include the consumer thread refactor (which will affect those 2 lines too). Also note that the race condition is not proven, but it looks fishy.